File: src/memento/control/proposals.py
from __future__ import annotations

import hashlib
import json
import sqlite3
from collections.abc import Iterable
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from enum import StrEnum
from typing import Any


class ProposalStatus(StrEnum):
    DRAFT = "draft"
    SUBMITTED = "submitted"
    APPROVED = "approved"
    REJECTED = "rejected"
    APPLIED = "applied"
    STALE = "stale"
    EXPIRED = "expired"


class ProposalTransitionError(RuntimeError):
    """Raised when a proposal state transition is invalid."""


@dataclass(frozen=True, slots=True)
class ProposalRecord:
    proposal_id: str
    author_principal: str
    client_instance_id: str | None
    base_revision: str
    intent: str
    rationale: str | None
    patch_json: str
    patch_hash: str
    status: ProposalStatus
    reviewed_by: str | None
    review_comment: str | None
    applied_operation_id: str | None
    applied_revision: str | None
    created_at: str
    updated_at: str
    expires_at: str | None

    @property
    def patch(self) -> dict[str, Any]:
        value = json.loads(self.patch_json)
        if not isinstance(value, dict):
            raise TypeError("proposal patch must decode to an object")
        return value


@dataclass(frozen=True, slots=True)
class ProposalAssetRecord:
    proposal_id: str
    asset_id: str
    concept_path: str
    asset_kind: str
    version: str
    media_type: str
    sha256: str
    blob_bytes: bytes
    manifest_json: str
    created_at: str

    @property
    def manifest(self) -> dict[str, Any]:
        value = json.loads(self.manifest_json)
        if not isinstance(value, dict):
            raise TypeError("proposal asset manifest must decode to an object")
        return value


@dataclass(frozen=True, slots=True)
class ProposalAssetInput:
    asset_id: str
    concept_path: str
    asset_kind: str
    version: str
    media_type: str
    sha256: str
    blob_bytes: bytes
    manifest_json: str


DEFAULT_PROPOSAL_TTL_DAYS = 30


def utcnow() -> str:
    return datetime.now(tz=UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def create_proposal(
    connection: sqlite3.Connection,
    *,
    proposal_id: str,
    author_principal: str,
    client_instance_id: str | None,
    base_revision: str,
    intent: str,
    rationale: str | None,
    patch: dict[str, Any],
    expires_in_days: int = DEFAULT_PROPOSAL_TTL_DAYS,
    assets: Iterable[ProposalAssetInput] = (),
) -> ProposalRecord:
    patch_json = json.dumps(patch, sort_keys=True)
    patch_hash = hashlib.sha256(patch_json.encode("utf-8")).hexdigest()
    now = utcnow()
    expires_at = (
        (datetime.now(tz=UTC).replace(microsecond=0) + timedelta(days=expires_in_days))
        .isoformat()
        .replace("+00:00", "Z")
    )
    asset_rows = tuple(assets)
    with connection:
        connection.execute(
            """
            INSERT INTO proposals(
                proposal_id, author_principal, client_instance_id, base_revision,
                intent, rationale, patch_json, patch_hash, status,
                created_at, updated_at, expires_at
            ) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                proposal_id,
                author_principal,
                client_instance_id,
                base_revision,
                intent,
                rationale,
                patch_json,
                patch_hash,
                ProposalStatus.SUBMITTED.value,
                now,
                now,
                expires_at,
            ),
        )
        _insert_proposal_assets(
            connection, proposal_id=proposal_id, assets=asset_rows, created_at=now
        )
    return get_proposal(connection, proposal_id)


def get_proposal(connection: sqlite3.Connection, proposal_id: str) -> ProposalRecord:
    row = connection.execute(
        "SELECT * FROM proposals WHERE proposal_id = ?", (proposal_id,)
    ).fetchone()
    if row is None:
        raise KeyError(proposal_id)
    return proposal_from_row(row)


def get_proposal_asset(
    connection: sqlite3.Connection, proposal_id: str, asset_id: str
) -> ProposalAssetRecord:
    row = connection.execute(
        "SELECT * FROM proposal_assets WHERE proposal_id = ? AND asset_id = ?",
        (proposal_id, asset_id),
    ).fetchone()
    if row is None:
        raise KeyError((proposal_id, asset_id))
    return proposal_asset_from_row(row)


def list_proposals(
    connection: sqlite3.Connection,
    *,
    status: ProposalStatus | None = None,
    author_principal: str | None = None,
) -> tuple[ProposalRecord, ...]:
    conditions: list[str] = []
    parameters: list[object] = []
    if status is not None:
        conditions.append("status = ?")
        parameters.append(status.value)
    if author_principal is not None:
        conditions.append("author_principal = ?")
        parameters.append(author_principal)
    where = f"WHERE {' AND '.join(conditions)}" if conditions else ""
    rows = connection.execute(
        f"SELECT * FROM proposals {where} ORDER BY created_at, proposal_id", parameters
    ).fetchall()
    return tuple(proposal_from_row(row) for row in rows)


def list_proposal_assets(
    connection: sqlite3.Connection,
    *,
    proposal_id: str | None = None,
    concept_path: str | None = None,
    asset_kind: str | None = None,
) -> tuple[ProposalAssetRecord, ...]:
    conditions: list[str] = []
    parameters: list[object] = []
    if proposal_id is not None:
        conditions.append("proposal_id = ?")
        parameters.append(proposal_id)
    if concept_path is not None:
        conditions.append("concept_path = ?")
        parameters.append(concept_path)
    if asset_kind is not None:
        conditions.append("asset_kind = ?")
        parameters.append(asset_kind)
    where = f"WHERE {' AND '.join(conditions)}" if conditions else ""
    rows = connection.execute(
        f"SELECT * FROM proposal_assets {where} ORDER BY created_at, proposal_id, asset_id",
        parameters,
    ).fetchall()
    return tuple(proposal_asset_from_row(row) for row in rows)


def update_proposal_status(
    connection: sqlite3.Connection,
    proposal_id: str,
    *,
    status: ProposalStatus,
    reviewed_by: str | None = None,
    review_comment: str | None = None,
    applied_operation_id: str | None = None,
    applied_revision: str | None = None,
) -> ProposalRecord:
    current = get_proposal(connection, proposal_id)
    assignments = ["status = ?", "updated_at = ?"]
    parameters: list[object] = [status.value, utcnow()]
    if reviewed_by is not None or current.reviewed_by is not None:
        assignments.append("reviewed_by = ?")
        parameters.append(reviewed_by)
    if review_comment is not None or current.review_comment is not None:
        assignments.append("review_comment = ?")
        parameters.append(review_comment)
    if applied_operation_id is not None or current.applied_operation_id is not None:
        assignments.append("applied_operation_id = ?")
        parameters.append(applied_operation_id)
    if applied_revision is not None or current.applied_revision is not None:
        assignments.append("applied_revision = ?")
        parameters.append(applied_revision)
    parameters.append(proposal_id)
    with connection:
        connection.execute(
            f"UPDATE proposals SET {', '.join(assignments)} WHERE proposal_id = ?",
            parameters,
        )
    return get_proposal(connection, proposal_id)


def proposal_from_row(row: sqlite3.Row) -> ProposalRecord:
    return ProposalRecord(
        proposal_id=row["proposal_id"],
        author_principal=row["author_principal"],
        client_instance_id=row["client_instance_id"],
        base_revision=row["base_revision"],
        intent=row["intent"],
        rationale=row["rationale"],
        patch_json=row["patch_json"],
        patch_hash=row["patch_hash"],
        status=ProposalStatus(row["status"]),
        reviewed_by=row["reviewed_by"],
        review_comment=row["review_comment"],
        applied_operation_id=row["applied_operation_id"],
        applied_revision=row["applied_revision"],
        created_at=row["created_at"],
        updated_at=row["updated_at"],
        expires_at=row["expires_at"],
    )


def proposal_asset_from_row(row: sqlite3.Row) -> ProposalAssetRecord:
    return ProposalAssetRecord(
        proposal_id=str(row["proposal_id"]),
        asset_id=str(row["asset_id"]),
        concept_path=str(row["concept_path"]),
        asset_kind=str(row["asset_kind"]),
        version=str(row["version"]),
        media_type=str(row["media_type"]),
        sha256=str(row["sha256"]),
        blob_bytes=bytes(row["blob_bytes"]),
        manifest_json=str(row["manifest_json"]),
        created_at=str(row["created_at"]),
    )


def _insert_proposal_assets(
    connection: sqlite3.Connection,
    *,
    proposal_id: str,
    assets: Iterable[ProposalAssetInput],
    created_at: str,
) -> None:
    for asset in assets:
        connection.execute(
            """
            INSERT INTO proposal_assets(
                proposal_id, asset_id, concept_path, asset_kind, version,
                media_type, sha256, blob_bytes, manifest_json, created_at
            ) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                proposal_id,
                asset.asset_id,
                asset.concept_path,
                asset.asset_kind,
                asset.version,
                asset.media_type,
                asset.sha256,
                asset.blob_bytes,
                asset.manifest_json,
                created_at,
            ),
        )

File: go/control/proposals.go
package control

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/rcarmo/memento/go/internal/pyjson"
)

type ProposalStatus string

const (
	Draft       ProposalStatus = "draft"
	Submitted   ProposalStatus = "submitted"
	Approved    ProposalStatus = "approved"
	Rejected    ProposalStatus = "rejected"
	Applied     ProposalStatus = "applied"
	Stale       ProposalStatus = "stale"
	NeedsRebase ProposalStatus = "needs_rebase"
	Conflicted  ProposalStatus = "conflicted"
	Expired     ProposalStatus = "expired"
)

func UnresolvedProposalStatuses() []ProposalStatus {
	return []ProposalStatus{Approved, Conflicted, Draft, NeedsRebase, Stale, Submitted}
}
func validProposalStatus(status ProposalStatus) bool {
	switch status {
	case Draft, Submitted, Approved, Rejected, Applied, Stale, NeedsRebase, Conflicted, Expired:
		return true
	}
	return false
}

type ProposalRecord struct {
	ProposalID         string         `json:"proposal_id"`
	AuthorPrincipal    string         `json:"author_principal"`
	ClientInstanceID   *string        `json:"client_instance_id"`
	BaseRevision       string         `json:"base_revision"`
	Intent             string         `json:"intent"`
	Rationale          *string        `json:"rationale"`
	PatchJSON          string         `json:"patch_json"`
	PatchHash          string         `json:"patch_hash"`
	Status             ProposalStatus `json:"status"`
	ReviewedBy         *string        `json:"reviewed_by"`
	ReviewComment      *string        `json:"review_comment"`
	AppliedOperationID *string        `json:"applied_operation_id"`
	AppliedRevision    *string        `json:"applied_revision"`
	CreatedAt          string         `json:"created_at"`
	UpdatedAt          string         `json:"updated_at"`
	ExpiresAt          *string        `json:"expires_at"`
}

func (r ProposalRecord) Patch() (map[string]any, error) {
	return objectJSON(r.PatchJSON, "proposal patch must decode to an object")
}

type ProposalAssetInput struct {
	AssetID      string `json:"asset_id"`
	ConceptPath  string `json:"concept_path"`
	AssetKind    string `json:"asset_kind"`
	Version      string `json:"version"`
	MediaType    string `json:"media_type"`
	SHA256       string `json:"sha256"`
	BlobBytes    []byte `json:"blob_bytes"`
	ManifestJSON string `json:"manifest_json"`
}
type ProposalAssetRecord struct {
	ProposalAssetInput
	ProposalID string `json:"proposal_id"`
	CreatedAt  string `json:"created_at"`
}

func (r ProposalAssetRecord) Manifest() (map[string]any, error) {
	return objectJSON(r.ManifestJSON, "proposal asset manifest must decode to an object")
}
func objectJSON(raw, message string) (map[string]any, error) {
	v, err := pyjson.Parse(raw)
	if err != nil {
		return nil, err
	}
	value, ok := v.(map[string]any)
	if !ok {
		return nil, errors.New(message)
	}
	return value, nil
}

type ProposalRequest struct {
	ProposalID, AuthorPrincipal, BaseRevision, Intent string
	ClientInstanceID, Rationale                       *string
	Patch                                             map[string]any
	ExpiresInDays                                     *int
	Assets                                            []ProposalAssetInput
}
type Proposals struct {
	DB  *sql.DB
	Now func() time.Time
}

func (p Proposals) now() time.Time {
	if p.Now != nil {
		return p.Now().UTC().Truncate(time.Second)
	}
	return time.Now().UTC().Truncate(time.Second)
}
func (p Proposals) Create(ctx context.Context, request ProposalRequest) (ProposalRecord, error) {
	var record ProposalRecord
	err := WithTransaction(ctx, p.DB, func(tx *sql.Tx) error { var err error; record, err = p.CreateInTx(ctx, tx, request); return err })
	if err != nil {
		return ProposalRecord{}, err
	}
	return record, nil
}

// CreateInTx is the explicit manage_transaction=False equivalent. Callers own
// commit/rollback, including atomic combination with access/idempotency events.
func (p Proposals) CreateInTx(ctx context.Context, tx *sql.Tx, request ProposalRequest) (ProposalRecord, error) {
	raw, err := pyjson.Dumps(request.Patch)
	if err != nil {
		return ProposalRecord{}, err
	}
	hash := fmt.Sprintf("%x", sha256.Sum256([]byte(raw)))
	now := p.now().Format(time.RFC3339)
	days := 30
	if request.ExpiresInDays != nil {
		days = *request.ExpiresInDays
	}
	if days < -3652058 || days > 3652058 {
		return ProposalRecord{}, errors.New("proposal expiry is out of range")
	}
	expiry := p.now().AddDate(0, 0, days)
	if expiry.Year() < 1 || expiry.Year() > 9999 {
		return ProposalRecord{}, errors.New("proposal expiry is out of range")
	}
	_, err = tx.ExecContext(ctx, `INSERT INTO proposals(proposal_id,author_principal,client_instance_id,base_revision,intent,rationale,patch_json,patch_hash,status,created_at,updated_at,expires_at) VALUES(?,?,?,?,?,?,?,?,?,?,?,?)`, request.ProposalID, request.AuthorPrincipal, request.ClientInstanceID, request.BaseRevision, request.Intent, request.Rationale, raw, hash, Submitted, now, now, expiry.Format(time.RFC3339))
	if err != nil {
		return ProposalRecord{}, err
	}
	for _, asset := range request.Assets {
		blob := asset.BlobBytes
		if blob == nil {
			blob = []byte{}
		}
		_, err = tx.ExecContext(ctx, `INSERT INTO proposal_assets(proposal_id,asset_id,concept_path,asset_kind,version,media_type,sha256,blob_bytes,manifest_json,created_at) VALUES(?,?,?,?,?,?,?,?,?,?)`, request.ProposalID, asset.AssetID, asset.ConceptPath, asset.AssetKind, asset.Version, asset.MediaType, asset.SHA256, blob, asset.ManifestJSON, now)
		if err != nil {
			return ProposalRecord{}, err
		}
	}
	return proposalByID(ctx, tx, request.ProposalID)
}

const proposalColumns = "proposal_id,author_principal,client_instance_id,base_revision,intent,rationale,patch_json,patch_hash,status,reviewed_by,review_comment,applied_operation_id,applied_revision,created_at,updated_at,expires_at"
const proposalAssetColumns = "proposal_id,asset_id,concept_path,asset_kind,version,media_type,sha256,blob_bytes,manifest_json,created_at"

type ProposalNotFoundError struct{ ProposalID, AssetID string }

func (e *ProposalNotFoundError) Error() string {
	if e.AssetID != "" {
		return "unknown proposal asset: " + e.ProposalID + "/" + e.AssetID
	}
	return "unknown proposal: " + e.ProposalID
}
func (p Proposals) Get(ctx context.Context, id string) (ProposalRecord, error) {
	return proposalByID(ctx, p.DB, id)
}

// GetInTx reads the updated proposal before the caller commits its review.
func (p Proposals) GetInTx(ctx context.Context, tx *sql.Tx, id string) (ProposalRecord, error) {
	return proposalByID(ctx, tx, id)
}
func proposalByID(ctx context.Context, db sqlExecutor, id string) (ProposalRecord, error) {
	record, err := scanProposal(db.QueryRowContext(ctx, "SELECT "+proposalColumns+" FROM proposals WHERE proposal_id=?", id))
	if errors.Is(err, sql.ErrNoRows) {
		return ProposalRecord{}, &ProposalNotFoundError{ProposalID: id}
	}
	return record, err
}
func scanProposal(row rowScanner) (ProposalRecord, error) {
	var r ProposalRecord
	err := row.Scan(&r.ProposalID, &r.AuthorPrincipal, &r.ClientInstanceID, &r.BaseRevision, &r.Intent, &r.Rationale, &r.PatchJSON, &r.PatchHash, &r.Status, &r.ReviewedBy, &r.ReviewComment, &r.AppliedOperationID, &r.AppliedRevision, &r.CreatedAt, &r.UpdatedAt, &r.ExpiresAt)
	if err != nil {
		return ProposalRecord{}, err
	}
	if !validProposalStatus(r.Status) {
		return ProposalRecord{}, fmt.Errorf("invalid proposal status: %s", r.Status)
	}
	return r, nil
}
func (p Proposals) GetAsset(ctx context.Context, id, assetID string) (ProposalAssetRecord, error) {
	record, err := scanProposalAsset(p.DB.QueryRowContext(ctx, "SELECT "+proposalAssetColumns+" FROM proposal_assets WHERE proposal_id=? AND asset_id=?", id, assetID))
	if errors.Is(err, sql.ErrNoRows) {
		return ProposalAssetRecord{}, &ProposalNotFoundError{ProposalID: id, AssetID: assetID}
	}
	return record, err
}
func scanProposalAsset(row rowScanner) (ProposalAssetRecord, error) {
	var r ProposalAssetRecord
	err := row.Scan(&r.ProposalID, &r.AssetID, &r.ConceptPath, &r.AssetKind, &r.Version, &r.MediaType, &r.SHA256, &r.BlobBytes, &r.ManifestJSON, &r.CreatedAt)
	return r, err
}

type ProposalQuery struct {
	Status               *ProposalStatus
	AuthorPrincipal      *string
	Unresolved           bool
	CurrentRevision, Now *string
	Limit                *int
	Cursor               *string
}

func (p Proposals) List(ctx context.Context, q ProposalQuery) ([]ProposalRecord, error) {
	conditions := []string{}
	args := []any{}
	if q.Status != nil {
		if q.CurrentRevision != nil && q.Now != nil {
			conditions = append(conditions, "(CASE WHEN expires_at IS NOT NULL AND expires_at < ? AND status != 'applied' THEN 'expired' ELSE status END) = ?")
			args = append(args, *q.Now, *q.Status)
		} else {
			conditions = append(conditions, "status = ?")
			args = append(args, *q.Status)
		}
	}
	if q.Unresolved {
		values := UnresolvedProposalStatuses()
		marks := []string{}
		for _, s := range values {
			marks = append(marks, "?")
			args = append(args, s)
		}
		conditions = append(conditions, "status IN ("+strings.Join(marks, ",")+")")
	}
	if q.Cursor != nil {
		var created, id string
		err := p.DB.QueryRowContext(ctx, "SELECT created_at,proposal_id FROM proposals WHERE proposal_id=?", *q.Cursor).Scan(&created, &id)
		if errors.Is(err, sql.ErrNoRows) {
			return nil, errors.New("invalid proposal list cursor")
		}
		if err != nil {
			return nil, err
		}
		conditions = append(conditions, "(created_at,proposal_id) > (?,?)")
		args = append(args, created, id)
	}
	if q.AuthorPrincipal != nil {
		conditions = append(conditions, "author_principal = ?")
		args = append(args, *q.AuthorPrincipal)
	}
	query := "SELECT " + proposalColumns + " FROM proposals" + whereClause(conditions) + " ORDER BY created_at, proposal_id"
	if q.Limit != nil {
		if *q.Limit < 1 {
			return nil, errors.New("proposal query limit must be positive")
		}
		query += " LIMIT ?"
		args = append(args, *q.Limit)
	}
	rows, err := p.DB.QueryContext(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanProposals(rows)
}
func whereClause(conditions []string) string {
	if len(conditions) == 0 {
		return ""
	}
	return " WHERE " + strings.Join(conditions, " AND ")
}
func scanProposals(rows operationRows) ([]ProposalRecord, error) {
	out := []ProposalRecord{}
	for rows.Next() {
		r, err := scanProposal(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, rows.Err()
}

type ProposalAssetQuery struct{ ProposalID, ConceptPath, AssetKind *string }

func (p Proposals) ListAssets(ctx context.Context, q ProposalAssetQuery) ([]ProposalAssetRecord, error) {
	return listProposalAssets(ctx, p.DB, q)
}
func (p Proposals) ListAssetsInTx(ctx context.Context, tx *sql.Tx, q ProposalAssetQuery) ([]ProposalAssetRecord, error) {
	return listProposalAssets(ctx, tx, q)
}

type proposalAssetQuerier interface {
	QueryContext(context.Context, string, ...any) (*sql.Rows, error)
}

func listProposalAssets(ctx context.Context, db proposalAssetQuerier, q ProposalAssetQuery) ([]ProposalAssetRecord, error) {
	conditions := []string{}
	args := []any{}
	for _, field := range []struct {
		name  string
		value *string
	}{{"proposal_id", q.ProposalID}, {"concept_path", q.ConceptPath}, {"asset_kind", q.AssetKind}} {
		if field.value != nil {
			conditions = append(conditions, field.name+" = ?")
			args = append(args, *field.value)
		}
	}
	rows, err := db.QueryContext(ctx, "SELECT "+proposalAssetColumns+" FROM proposal_assets"+whereClause(conditions)+" ORDER BY created_at, proposal_id, asset_id", args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanProposalAssets(rows)
}
func scanProposalAssets(rows operationRows) ([]ProposalAssetRecord, error) {
	out := []ProposalAssetRecord{}
	for rows.Next() {
		r, err := scanProposalAsset(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, rows.Err()
}

// UpdateStatus is the low-level blind update, not a review/apply policy check.
// Omitting optional review/application fields clears their prior values.
type ProposalStatusUpdate struct {
	Status                                                         ProposalStatus
	ReviewedBy, ReviewComment, AppliedOperationID, AppliedRevision *string
}

func (p Proposals) UpdateStatus(ctx context.Context, id string, update ProposalStatusUpdate) (ProposalRecord, error) {
	if !validProposalStatus(update.Status) {
		return ProposalRecord{}, fmt.Errorf("invalid proposal status: %s", update.Status)
	}
	current, err := p.Get(ctx, id)
	if err != nil {
		return ProposalRecord{}, err
	}
	assignments := []string{"status=?", "updated_at=?"}
	args := []any{update.Status, p.now().Format(time.RFC3339)}
	for _, field := range []struct {
		name       string
		value, old *string
	}{{"reviewed_by", update.ReviewedBy, current.ReviewedBy}, {"review_comment", update.ReviewComment, current.ReviewComment}, {"applied_operation_id", update.AppliedOperationID, current.AppliedOperationID}, {"applied_revision", update.AppliedRevision, current.AppliedRevision}} {
		if field.value != nil || field.old != nil {
			assignments = append(assignments, field.name+"=?")
			args = append(args, field.value)
		}
	}
	args = append(args, id)
	if _, err = p.DB.ExecContext(ctx, "UPDATE proposals SET "+strings.Join(assignments, ",")+" WHERE proposal_id=?", args...); err != nil {
		return ProposalRecord{}, err
	}
	return p.Get(ctx, id)
}
